### PR TITLE
Add more generic 'Media' object to posts model (Open to comments)

### PR DESCRIPTION
**Pull request by picklechips**
The idea with this PR is to generalise media types on posts.  "Media" meaning images, polls, videos, etc.

This allows our types of posts to be much more easily extendable, since all you have to do is create a model for the new type and add it into the `Media` model.

As an example, in the future if we want to add polls to posts we would just create a "Poll" model, and then add the following code snippet into the Media model:

```
poll: {
	type: Schema.Types.ObjectId,
	ref: 'Poll',
	validate: mediaValidator.bind(this)('poll')
}
```

Add a wee bit of creation logic, and then you're done. Wouldn't even have to touch the Posts model.

Then on the client-side, we could simply check the type of media on the post and then display it appropriately.

You can check out the tests to get somewhat of an idea of the usage. Looking for thoughts or opinions, curious as to what others think.

### CODE REVIEW

File: models/Media/Media.js [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+MediaSchema.statics.create = function (type, data) {
+	return new Promise((resolve, reject) => {
+		if (type === 'image') {
+			sharp(data)
+			.resize({ height: 600, width: 600, withoutEnlargement: true })
+			.jpeg()
+			.toBuffer()
+			.then(outputBuffer => {
+				const newImage = new PostPicture({ buffer: outputBuffer });
+				newImage.save().then(pic => {
```
Comment (picklechips) [inline models/Media/Media.js]: Come to think of it, all this creation logic should really be in the `PostPicture` model.

**Comment by scholvat** [comment]
Seems like a good organization structure, though getting this to be backwards compatible might be a challenge. I would think a script would also need to be written to convert all posts to this new schema.

**Comment by picklechips** [comment]
It would, but I wouldn't worry about that until after it's all merged in. I think the correct way about doing this is the following:

- Make the new changes on the server leaving in the legacy code
- Update the client to handle both legacy & new posts
- Migrate the legacy posts to new posts
- Remove legacy code from server & client

File: models/Media/Media.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+		ref: 'PostPicture',
+		validate: mediaValidator.bind(this)('image')
+	}
+	// TODO: poll
+}, { timestamps: true });
+
+MediaSchema.statics.create = function (type, data) {
+	return new Promise((resolve, reject) => {
+		if (type === 'image') {
+			PostPicture.create(data).then(image => {
```
Comment (aopal) [inline models/Media/Media.js]: Can you please put this in a helper function?

File: models/Media/PostPicture.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+	buffer: {
+		type: Buffer,
+		required: true,
+	},
+});
+
+PostPictureSchema.statics.create = function (buffer) {
+	return new Promise((resolve, reject) => {
+		sharp(buffer)
+		.resize({ height: 600, width: 600, withoutEnlargement: true })
```
Comment (aopal) [inline models/Media/PostPicture.js]: Why are we resizing? Does a non-square image cause issues in the UI?

Comment (picklechips) [reply models/Media/PostPicture.js]: This was introduced as part of #86. Here I'm just moving the logic from the `post` model to the `postPicture` model.

I had very similar questions in that PR which were resolved over a private chat. Essentially though, it's just to reduce file size.

Although looking it at it here, I think this is supposed to be using the [`fit inside`](https://sharp.dimens.io/en/stable/api-resize/#resize) option in order to maintain aspect ratio...

@scholvat can you verify this? Its being used in your scripts but not in the actual code.

Comment (scholvat) [reply models/Media/PostPicture.js]: Yep that’s a mistake. Feel free to fix it in this PR or make a separate patch to update that. It should match what is in the scripts.

**Comment by picklechips** [comment]
This has been updated. @scholvat @aopal 

I've also opened another PR for the polls model, which demonstrates how this media object would be used.

https://github.com/picklechips/skybunk-server/pull/1

**Comment by picklechips** [comment]
@scholvat @aopal @neilbrub 
I'll be merging this tomorrow sometime if there's no other comments!

**Pull request merged**